### PR TITLE
Initial concept implementation

### PR DESCRIPTION
**Pull request by gorbuvla**
Just a concept implementation, no CI, no bintray publishing

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    dependencies {
+        classpath 'com.android.tools.build:gradle:3.3.2'
+        classpath "org.jetbrains.kotlin:kotlin-gradle-plugin:$kotlin_version"
+    }
+}
+
+apply plugin: 'java-library'
+apply plugin: 'org.jetbrains.kotlin.jvm'
+
+allprojects {
```
Comment (davidbilik) [inline build.gradle]: probably unnecessary

File: build.gradle [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+repositories {
+    google()
+    jcenter()
+    maven { url "https://jitpack.io" }
+}
+
+dependencies {
+    implementation "com.squareup.retrofit2:retrofit:2.5.0"
+    implementation 'org.jetbrains.kotlin:kotlin-stdlib:1.3.21'
+    implementation 'org.jetbrains.kotlinx:kotlinx-coroutines-core:1.2.0-alpha'
```
Comment (davidbilik) [inline build.gradle:26]: I think there is newer version of coroutines, isn it?

**Comment by davidbilik** [comment]
I think that the concept is great, just like I imagined. Are you satisfied with this solution?

**Comment by gorbuvla** [comment]
@davidbilik main concern for me was "HOT" `Deferred` in original adapterFactory by JW. Can't imagine a real situation where it could brake things down, but this eager execution seems sort of unnatural. I would expect execution to start whenever `.await()` is called in some coroutine scope, not when an instance of `Deferred` is obtained.

Chaining could be done much cooler by just using `.fold`, but i guess it is up to `Interceptor` whether to continue in chain, couldn't figure out a cool functional approach to it...

**Pull request merged**